src/refs/mod.rs

```diff
@@ -137,74 +199,60 @@ impl Token {
                 // we handle the first element separately, so we can establish a local mutable
                 // variable which we can update during the walk of the parameters Mapping.
                 let k0 = refpath_iter.next().unwrap();
-                let k0: Value = k0.into();
                 // v is the value which we update to point to the next value as we recursively
                 // descend into the params Mapping
                 let mut v = params
-                    .get(&k0)
-                    .ok_or_else(|| anyhow!("[k0] unable to lookup key '{k0}' for '{path}'"))?;
+                    .get(&k0.into())
+                    .ok_or_else(|| state.render_missing_key_error(&path, k0))?;
 
                 // newv is used to hold temporary Values generated by interpolating v
                 let mut newv;
 
                 // traversed is used to keep track of the path segments we've already processed
-                let mut traversed = vec![k0.as_str().unwrap().to_string()];
+                let mut traversed = vec![k0.to_string()];
 
                 // descend into the params Mapping, looking up each segment of the reference path
                 // sequentially, updating `v` and `newv` as we go.
                 for key in refpath_iter {
-                    match v {
+                    // For lookups into Strings and ValueLists, we locally interpolate the
+                    // value into `newv` so we don't have to worry about the order in which
+                    // individual references are resolved, and always do value lookups on
+                    // resolved references.
+                    newv = interpolate_string_or_valuelist(v, params, state)?;
```

Comment:
Note that I compared the performance of always interpolating (or cloning) the value with the old approach where we don't clone non-String/non-ValueList values and didn't find any significant performance difference for this change.

Since the new approach makes `Ref::resolve()` easier to understand and reduces the amount of duplication for error handling, I've opted to keep this change.